optimum/neuron/distributed/base.py

```diff
@@ -622,15 +625,39 @@ def should_parallelize_layer_predicate_func(layer):
                 # It means that `_MODEL_PARALLEL_RNG_TRACKER_NAME` was already added to the rng tracker, we can ignore.
                 pass
 
-            model = cls._parallelize(
-                model,
+            tied_parameters = get_tied_parameters_dict(model)
+
+            cls._parallelize(
+                orig_model,
                 device=device,
                 parallelize_embeddings=parallelize_embeddings,
                 sequence_parallel_enabled=sequence_parallel_enabled,
                 should_parallelize_layer_predicate_func=should_parallelize_layer_predicate_func,
                 skip_linear_weight_load=skip_linear_weight_load,
                 kv_size_multiplier=kv_size_multiplier,
             )
+
+            for param_name, root_tied_param_name in tied_parameters.items():
+                parent_mod, attr_name = get_parent_module_and_param_name_from_fully_qualified_name(model, param_name)
+                param = getattr(parent_mod, attr_name)
+                root_parent_mod, root_attr_name = get_parent_module_and_param_name_from_fully_qualified_name(
+                    model, root_tied_param_name
+                )
+                root_tied_param = getattr(root_parent_mod, root_attr_name)
+                if getattr(param, "tensor_model_parallel", False) and not getattr(
+                    root_tied_param, "tensor_model_parallel", False
+                ):
+                    # In this case it means that `param` was parallelized but not `root_tied_param`.
+                    # It will be overwritten by root_tied_param when tiying the weights if we do not do anything.
+                    # We tie `root_tied_param` to `param`.
+                    # What will happen is as follows:
+                    #   - If weight_map contains a checkpoint for `param_name`, it has already been initialized or will
+                    #   be initialized with `cls._maybe_load_weights_to_parallel_linear`.
+                    #   - Otherwise, it has not been initialized and `cls._initialize_or_load_weights` will take care
+                    #   of it.
+                    root_parent_base_mod, _ = get_base_model_and_peft_prefix(root_parent_mod)
+                    setattr(root_parent_base_mod, root_attr_name, param)
```

Comment:
It is a bit unclear why setting the param in the base model (and not only in the peft model) will ultimately triggers its parallelization.

Reply:
Basically the peft model contains the base model. They have the same exactly the same parameters.
Here what we are trying to solve is this:

1. We have parameters that are tied before parallelization, meaning they share the same weight. 
2. One of these weight is parallelized during parallelization, meaning that a new parameter is created to replace it.
3. The other parameter has not been replaced so we end-up with two parameters: one sharded and one full weight. 
4. We have a method to re-tie the weights after parallelization / moving the weights to the XLA device. If this method uses the full weight as the "reference" instead of the parallelized one it will overwrite it even though that is the one we actually wanted to keep. 
5. To prevent that, we check all the tied weights right after parallelization. If one of them is left unchanged while the other has been sharded, we updated it to the sharded version.

Reply:
The peft model does not contain other parameters than the base model. So when I overwrite the parameter, I need to parent module of the parameter. It might be not useful here to `get_base_model_and_peft_prefix` here (I do not remember which occasion brought me to do that) but it's a no-op if the model is already a base model.